[#48] UserService 및 관련 도메인 코틀린 전환

msugo1 opened the pull request:
## 📖  의존성 관련

### 📑   pom.xml

- 기본 클래스 속성이 상속 불가능한 final인 코틀린이 필요한 경우 open 키워드를 붙여주는 `kotlin-allopen` 의존성 추가

## 📖 Model 관련

### 📑 User.kt

- 기존 UserService 내부에 갖고 있던 private 메소드를 User 내부의 동반객체 메소드로 변경하여 사용자 정보 검증로직을 User 클래스가 담당하도록 변경

- 비밀번호와 이메일을 검증하는 각각의 메소드 `verifyPassword`, `verifyEmail` 메소드 추가

### 📑 LoginData.kt (<- UserLoginData)

- DB에서 로그인 요청을 한 유저의 정보를 얻어온 뒤 이를 담은 객체를 만들기 위해 사용되는 클래스 (코틀린으로 전환)

### 📑 FileData.kt (<- FileData.java)

- 프로필 이미지 정보를 담은 객체를 생성하기 위한 클래스

### 📑 Email.kt (<- BaseEmail, FindPasswordEmail)

- 동반객체 내부에 `create` 메소드를 만들어 이메일 내용 작성 시 java의 static 메소드처럼 활용할 수 있도록 구성 (Email 객체 리턴)

- 최상위 층에 const 상수를 두어 기존처럼 상속을 활용하지 않아도 각 이메일에 맞는 Title과 Body를 가질 수 있도록 구성

## 📖 Service 관련

### 📑 UserService.kt

- 기존 UserService.java를 UserService.kt로 리팩토링

## 📖 Mapper 관련

### 📑 UserMapper.java

- 코틀린은 `기본적으로 null을 허용하지 않고`, `null을 안전하게 처리할 ?: (엘비스 연산자)`를 가지고 있으므로 리턴 값에 적용된 Optional 제거

## 📖 리팩토링 시 코드 수정내용 소스코드와 테스트 코드에 반영

Commit:
[#48 UserService 및 관련 도메인 코틀린 전환

* pom.xml

- 기본 클래스 속성이 상속 불가능한 final인 코틀린이 필요한 경우 open
  키워드를 붙여주는 `kotlin-allopen` 의존성 추가

Model 관련

* User.kt

- 기존 UserService 내부에 갖고 있던 private 메소드를 User 내부의 동반
  객체 메소드로 변경하여 User과 관련있는 로직을 처리하고 있음을 두각
  시키도록 변경

- 비밀번호와 이메일을 검증하는 각각의 메소드 `verifyPassword`,
  `verifyEmail` 메소드 추가

* LoginResponse.kt (<- UserLoginData)

- DB에서 로그인 요청을 한 유저의 정보를 얻어온 뒤 이를 담은 객체를
  만들기 위해 사용되는 클래스 (코틀린으로 전환)

* FileData.kt (FileData.java)

- 프로필 이미지 정보를 담은 객체를 생성하기 위한 클래스

Service 관련

* UserService.kt

- 기존 UserService.java를 UserService.kt로 리팩토링

Mapper 관련

* UserMapper.java

- 코틀린은 `기본적으로 null을 허용하지 않고`, `null을 안전하게 처리할 ?:
 (엘비스 연산자)`를 가지고 있으므로 로그인 시 로그인 관련 데이터를
 반납해 줄 때, Optional 제거

테스트 관련

- 코틀린 전환 후 테스트가 깨진 부분 해결

Commit:
[#48] UserService 코틀린 전환 마무리

Model 관련

* BaseEmail, FindPasswordEmail

- Email.kt 하나로 통합

- 동반객체 내부에 `create` 메소드를 만들어 이메일 내용 작성 시 java의 static 메소드처럼 활용할 수 있도록 구성 (Email 객체 리턴)

- 최상위 층에 const 상수를 두어 각 이메일에 맞는 Title과 Body를 가질 수
  있도록 구성

Service 관련

* UserService

- 사용자의 비밀번호를 찾는 `findPassword` 메소드와 로그인 정보를
  얻어오는 `getValidUserLoginData` 메소드의 리팩토링 마무리

리팩토링 시 코드 수정내용 소스코드와 테스트 코드에 반영

f-lab-dev commented on src/main/java/me/soo/helloworld/model/user/User.kt:
```diff
@@ -44,6 +46,18 @@ data class User @JvmOverloads constructor(
 ) {
 
+    companion object {
+        fun verifyPassword(reqPassword: String, userPassword: String, passwordEncoder: PasswordEncoder) {
```
이건 외부 클래스로 분리하거나, 내부에서만 쓰는거면 private으로 해주는게 낫지 않을까요?

msugo1 replied:
UserService 내부에서만 사용하기 때문에 UserService로 다시 옮기고, private으로 선언해 주었습니다.

Commit:
[#48] 피드백 반영

* User.kt 내 동반 객체 메소드를 UserService private으로 다시 변경

- UserService 내부에서만 활용되는 `verifyPassword`, `verifyEmail`
  메소드를 다시 UserService 내부에 위치 시킨 후 private으로 변경

- verifyEmail의 경우는 한 곳에서만 쓰이므로, 분리한 메소드의 내부를 다시
  원 메소드로 위치시킴
